Platform runtime revision | v0.3.0 r8

AndroDevcd opened the pull request.

AndroDevcd committed:
Initial commit to rebuild runtime system

AndroDevcd committed:
updated VM

AndroDevcd committed:
got new task scheduling system to work

AndroDevcd committed:
fixed issues within task scheduling system:
- Task scheduling system is now stable

AndroDevcd committed:
added initial code for compiler frontend

AndroDevcd committed:
added json settings processing

AndroDevcd committed:
initial commit for task delegator system

AndroDevcd committed:
added support for scheduling tasks

AndroDevcd committed:
Fixed small bug with task delegation system

AndroDevcd committed:
Initial code for class pre processing

AndroDevcd committed:
added initializer param support for init decl

AndroDevcd committed:
added initial code for type matching

AndroDevcd committed:
added initial code for creating constructors

AndroDevcd committed:
fixed small bugs with type matching and function resolution

AndroDevcd committed:
added initial code for class pre-processing

AndroDevcd committed:
added initial code for creating aliases

AndroDevcd committed:
added initial code for pre-processing fields

AndroDevcd committed:
Added initial support for:
  - New Import syntax & import groups
  - Elvis operator
  - Nullables
  - Base class specifier for generics
  - Paramaterized array access via operator overload

AndroDevcd committed:
Added initial code for caching

AndroDevcd committed:
spead up tokenization process

AndroDevcd committed:
added initial code to process types

AndroDevcd committed:
added utype parsing

AndroDevcd committed:
added initial code for resolving a single item

AndroDevcd committed:
Finished normal item resolving

AndroDevcd committed:
added support to resolve function pointer types

AndroDevcd committed:
added support to resolve function calls

AndroDevcd committed:
added initial code for resolving anything in the compiler

AndroDevcd committed:
added support for processing base classes

AndroDevcd committed:
added support for processing interfaces

AndroDevcd committed:
added the following:
- Support for base class checking for resolving fields
- Support for nullable access
- Support for post-processing fields & aliases
- Initial code for processing expressions

AndroDevcd committed:
added support for post-processing methods, delegates, operators, constructors, and initializers

AndroDevcd committed:
Added support for the following:
- Calling functions via field type
- Compiling function return types on demand

AndroDevcd committed:
added support for processing largest file first

AndroDevcd committed:
fixed all compiler bugs

AndroDevcd commented:
Compiler is now 50% done!

AndroDevcd committed:
Fixed bugs w/ processing delegates, resolving functions, and more

AndroDevcd committed:
small bug fix, expression compilation started

AndroDevcd committed:
added initial support for processing minus expression

AndroDevcd committed:
added support for processing minus expression

AndroDevcd committed:
added support for processing self & dot notation cal expressions & added several checks for nullables & illegal property access

AndroDevcd committed:
Fixed bugs with the following:
- Resolving functions
- Self expression
- operation struct copying

Added support for the following:
- Dynamically calling resolve with starting class provided

AndroDevcd committed:
Update ROADMAP.md

AndroDevcd commented:
Compiler expression dev paused to include a built-in dependency injection system (this will be a major feature + selling point of the language.

AndroDevcd committed:
Added initial code for dependency injection system

AndroDevcd committed:
Nearly finished code for dependency injection system

AndroDevcd committed:
Added support for dependency injection system & fixed several bugs

AndroDevcd committed:
modified how no named components work

AndroDevcd commented:
Compiler dev resumed dependency injection framework is finished

AndroDevcd committed:
fixed and improved dependency injection framework, new expression compilation dev continuing

AndroDevcd committed:
Further fixed bugs in the compiler & finished new expression compiling

AndroDevcd committed:
Fixed remaining outstanding bugs on compiler expression compiling continuing

AndroDevcd committed:
Updated readme

AndroDevcd committed:
Update README.md

AndroDevcd committed:
Update README.md

AndroDevcd committed:
Update README.md

AndroDevcd committed:
Added explicit type requirement feature for get() keyword

AndroDevcd committed:
Added initial code to compile lambdas

AndroDevcd committed:
Added initial support for the following:
- Sizeof expressions
- Paren expressions
- Post increment expressions
- Cast expressions

AndroDevcd committed:
Added support for post Ast expressions

AndroDevcd committed:
Tested and fixed bug with expression compilation

AndroDevcd committed:
Fixed remaining compiler bugs

AndroDevcd committed:
Primary expression compilation finished

AndroDevcd committed:
Added support for dictionary expressions

AndroDevcd committed:
Added the following:
- Support for elvis operator expressions
- Support for dynamic arrays

AndroDevcd commented:
Compiler expression dev paused to add in generic functions

AndroDevcd committed:
Added the following:
- Support for &, ^, |, &&, || binary expressions
- Initial Support for = binary expressions

AndroDevcd committed:
Added missing test files

AndroDevcd committed:
Added the following:
- Updated elvis operator to support differing types
- Added support for and expression
- Added initial support for assign expressions

AndroDevcd committed:
Added the following:

- Added support for +, -, *, /, %, <<, >>, ==, !=, <, >, <=, >=, and ** expression operations
- Updated task scheduling system in virtual machine
- Fixed several bugs within the compiler

AndroDevcd committed:
Added the following:
- Fixed bug where inner fields, functions, and enums could not be accessed via "self->"
- Fixed ambiguous complaint in std lib
- Updated resolve system to support filtering out functions

AndroDevcd committed:
Added the following:
- Added support for =, +=, -=, %=, *=, ^=, &=, /= expressions

AndroDevcd committed:
Added the following:
- Added support for =, +=, -=, %=, *=, ^=, &=, /=, and |= expressions

AndroDevcd committed:
Merge remote-tracking branch 'origin/runtime-3.0.0_r8' into runtime-3.0.0_r8

AndroDevcd committed:
Added the following:
- Added initial support for init declarations
- Compilation stopped if not all files are tokenized and parsed
- Removed support for resolving base constructors without correct name
- Modified how the final compilation stage works to better support generic classes
- Added initial support for compiling functions

AndroDevcd commented:
Nearly finished with the compiler, In the final stages of compilation. Code generation will be the only thing left after that!!

AndroDevcd committed:
Added the following:
- Fixed bug with local field resolution
- We worked how blocks are processed to allow for greater support for catching user errors in statements
- Added support for "return" statements

AndroDevcd committed:
Added the following:
- Added support for "If" statements
- Added support for labels in code

AndroDevcd committed:
Added the following:
- Added support for "local variable", "local alias", "for", "foreach", "while", "dowhile", "throw", "continue", "break", "lock", "trycatch", "when", and "Label" statements
- Added support for labels in code

AndroDevcd committed:
Added the following:
- Added support for "local variable", "local alias", "for", "foreach", "while", "dowhile", "throw", "continue", "break", "lock", "trycatch", "when", and "Label" statements
- Added support for labels in code
- Fixed major bug in parser for processing types

AndroDevcd committed:
Added the following:
- Fied most bugs with statements
- Fixed a lot of compiler bugs

AndroDevcd committed:
Added the following:
- Fixed more broken statements
- Added `!!` postfix expression
- Fixed parse error bugs
- Updated a few std lib files to support new compiler
- Fixed nasty race condition for function return type compiling

AndroDevcd commented:
The compilation portion is finished!! (For the most part)
There are lots of unfinished parts in the compiler but the foundation and skeleton is there. Moving on to the code generation/Optimization part and will fix bugs and gaps in logic on an as-needed basis

AndroDevcd committed:
Added the following:
- Added initial support for code generation
- Added initial support for optimizer

AndroDevcd committed:
Added the following:
- Added initial support for "++", "--", "!", and "-" prefix operations
- Added support for allocating internal registers
- Foundation for code generator is 40-50% done!

AndroDevcd committed:
Added the following:
- Added support for creating new classes
- Added support for calling functions

AndroDevcd committed:
Added the following:
- Too much to track, I added a shitload of support for the code generator and fixed a crap ton of bugs

AndroDevcd committed:
Added the following:
- All expressions can now be generated to code (for the most part) there may still be some tiny gaps left. Will start full expression testing soon

AndroDevcd committed:
Added the following:
- added supportfor post "++"

AndroDevcd commented:
All expressions can now be generated to the VM machine code (for the most part). Now will be running tests on all expressions and moving on to the final code generation stage which is processing for loops, if statements, etc.

Compiler is 70 - 80% done

AndroDevcd committed:
Added the following:
- expression code generation done!

AndroDevcd committed:
Added the following:
- Updated dependency resolution system to be more efficient and fully capture all dependencies

AndroDevcd committed:
Added the following:
- Expressions have been tested!

dezashibi commented:
Hi @AndroDevcd
I'm interested to know about the development status of the compiler, if contibution is opened or not

AndroDevcd committed:
Added the following:
- Added support for "for", "local variable", "if", "foreach", and "When" statements code generation

AndroDevcd commented:
> Hi @AndroDevcd I'm interested to know about the development status of the compiler, if contibution is opened or not

Hey @knavels ! First off thanks for being interested in the status and contributing to the language :). Currently unfortunately due to the overall complexity of the compiler, this PR and the language is not open for contributions. However, shortly after this PR is merged and I deploy the next version **Sharp v0.3.0 r8**. I am planning on ramping up community efforts for the project.

As for the status of the overall project, here's a snapshot of where things are:
- Currently, this PR serves as the first official "stable" progr. language release. Over the years it's been ridden with frustrating nasty bugs and a lack of features. This PR solves a lot of those issues as well as adds a ton of features to the language.
- As of today **8/22/2022 11PM CST** I am very close to the end of development with the compiler. I am currently building out the code generator to actually convert the sharp code to my VM's machine language. Once that is done here's a rough but non-exhaustive list of what else needs to be done:
    - Finish code generator to support all statements (should be finished by the end of this week)
    - Finish adding support for closures in the compiler (this is relatively trivial to finish up few days tops)
    - Plug any additional logic gaps and holes in the compiler (this will consist of resolving all the TODO's in the code)
    - Add support for code obfuscation complexity (? not sure)
    - Add in support for file caching (this will be very challenging)
         - File caching should speed up the overall compilation time by at minimum 50% as currently, it takes a min. of **5.5s** to compile everything
    - Add in support for running the compiler in "watch mode" which allows live updating of your compiled code. Basically, as you write code your sharp exe file will be updated in real-time so you don't have to keep recompiling everything.
    - Run tests on the compiled code on the runtime system.
    - Add support for optimizing sharp code. (very challenging may take up to 1-2months to complete)
    - Rewrite entire standard library (this sounds like a big lift but is INFINITELY easier than all the things listed above combined) this will be the final thing needed to complete this PR.

So in short although I am close to the "end" of compiler development. There's clearly a bit of work left to be done. I'm hoping that by this December to have all this done. That way I can start back on youtube and give an update on everything. I also have plans for creating a discord server and more! So definitely lots of good things coming in the near future!!

dezashibi commented:
@AndroDevcd I'm glad that this project is under development, this area (Compilers and Programming languages) has been always my passion and favorite and I can't describe by words, but I wish all goes as you've planned. also looking forward to have you back on youtube

AndroDevcd committed:
Added the following:
- Added support for "try" statements code generation
- Added support for official finally block execution when returning out of a function

AndroDevcd committed:
Added the following:
- Added support for "try", "do-while", "while", "continue", "break", "return", and "lock" statements code generation
- Added support for official finally block execution when returning out of a function

AndroDevcd commented:
@knavels I'm right there with you it's my passion as well, also thanks for the well wishes. See you on the youtube channel in a few months!

AndroDevcd committed:
Added the following:
- Fixed ALL bugs with code generation. All functions can now be compiled

AndroDevcd committed:
Added the following:
- Updated "&&" binary operation to work correctly
- Fixed minor bugs
- Added explicit "override" requirement for delegate functions
- Fixed massive error reporting bugs specifically with generic classes

AndroDevcd committed:
Added the following:
- Fixed massive bugs
- Added "!!", "!!.", and "?." support when accessing fields and objects
- Added debugging code for the operation_schema system
- Updated how constructors are initialized

AndroDevcd committed:
Added the following:
- Added support for "asm" statement

AndroDevcd committed:
Added the following:
- Added initial support for code obfuscation

AndroDevcd committed:
Added the following:
- Started std lib rewrite

AndroDevcd committed:
Added the following:
- Partially fixed issues with generic classes (broken code)

AndroDevcd committed:
Added the following:
- Fixed all known bugs in compiler (Gearing up for runtime execution testing now)

AndroDevcd committed:
Added the following:
- Minor bug fixes made

AndroDevcd committed:
Added the following:
- Minor bug fixes made
- Fixed issue with asm statements processing code wrong

AndroDevcd committed:
Added the following:
- Added support for exe file generation

AndroDevcd committed:
Added the following:
- Added support for exe file generation
- bunch of other stuff

AndroDevcd committed:
Added the following:
- Locking and unlocking objects
- New system for context swtching fibers
- Added idle scheduler for more efficent taskk management
- Finished garbage collector
- Partially cleaned up exception system

AndroDevcd committed:
Added the following:
- Fixed thredding system
- updated main file
- nearly done with vm rewrite

AndroDevcd committed:
Added the following:
- finished exception system
- finished vm main loop

AndroDevcd committed:
Added the following:
- added new serialization/deserialization system
- Initial support for system interrupts

AndroDevcd committed:
Added the following:
- got vm to build

AndroDevcd committed:
Added the following:
- fixed compiler bugs
- added obfuscator modification

AndroDevcd committed:
Added the following:
- Fixed recursive loop with getters and setters

AndroDevcd committed:
Added the following:
- Fixed recursive loop with getters and setters
- A crap ton more stuff

AndroDevcd committed:
Added the following:
- Fix a crap ton of stuff
- WE GOT TO RUN THE FIRST EVER HELLO WORLD!!!!!!

AndroDevcd commented:
First ever hello world ran in the language with the new system!!! Its been a long time coming. Time to test everything and get it ready to be released!

AndroDevcd committed:
Added the following:
- Fix a crap ton more stuff
- testing continued

AndroDevcd committed:
Added the following:
- Fix a crap ton more stuff
- MASSIVE changes to the dependency injection system

AndroDevcd committed:
Added the following:
- Fix a crap ton more stuff
- Ready to begin code optimization phase

AndroDevcd committed:
Added the following:
- Fixed too much crap to remember!
-Added c++ support

AndroDevcd committed:
Added the following:
- removed unessicary code

AndroDevcd committed:
Added the following:
- Fixed compiler issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed runtime issues for linux

AndroDevcd committed:
Added the following:
- Fixed file system code issue

AndroDevcd committed:
Added the following:
- Fixed native function setup bug

AndroDevcd committed:
Added the following:
- Fixed issue that crashes compiler for unused native functions

AndroDevcd committed:
Added the following:
- Fixed issue that crashes compiler for unused native functions

AndroDevcd committed:
Added the following:
- Fixed issues with scheduler and c++ calls

AndroDevcd committed:
Added the following:
- Fixed issues with scheduler and c++ calls

AndroDevcd committed:
Added the following:
- Fixed issues with scheduler and c++ calls

AndroDevcd committed:
Added the following:
- Fixed issues with scheduler and c++ calls

AndroDevcd committed:
Added the following:
- Fixed issues with scheduler and c++ calls

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Fixed more issues

AndroDevcd committed:
Added the following:
- Massive fixes made

AndroDevcd committed:
Added the following:
- Massive fixes made

AndroDevcd committed:
Added the following:
- Massive fixes made

AndroDevcd committed:
Added the following:
- Massive fixes made

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- More bug fixes

AndroDevcd committed:
Added the following:
- Fixed massive issue with double scheduling of tasks
- Fixed string $ bug
- Removed compiler options
- modified build class in runtime system

AndroDevcd committed:
Added the following:
- commiting obfuscation code

AndroDevcd committed:
Added the following:
- Created official build script command
- Added code obfuscation support
- Added standard library class requirements

AndroDevcd committed:
Added the following:
- Minor bug fixes
- initial pass on non-nullable field initialization enforcement (unfinished)

AndroDevcd committed:
Added the following:
- variable instantation

AndroDevcd committed:
Added final touches to compiler

AndroDevcd committed:
updated build.sh

AndroDevcd committed:
fixed serialization bug

AndroDevcd committed:
sped up object serialization

AndroDevcd committed:
update README.md

AndroDevcd committed:
Bug fixes and improvements

AndroDevcd committed:
Merge remote-tracking branch 'refs/remotes/origin/runtime-3.0.0_r8' into runtime-3.0.0_r8

AndroDevcd committed:
Improved serialization performance

AndroDevcd committed:
Continued compiler bootstrapping

AndroDevcd committed:
Updated serialization performance

AndroDevcd committed:
Updated serialization + added reflection

Commit:
Making progress on the jit

Commit:
Finished jit sections + added instructions

Commit:
Added opcodes

Commit:
Added stack frame checking

Commit:
updated build file

AndroDevcd committed:
updated cmake

AndroDevcd committed:
Added initial instructions

AndroDevcd committed:
Updated platform arch checks

Commit:
fixed small code bugs

Commit:
attemot to fix multi function compile

Commit:
Fixed jit mem corruption bug
